### DISCUSSION TITLE
File size blows up for some files

### DISCUSSION THREAD

**Discussion by zain910128** [py-pdf/pypdf#2699] [Q&A] [answered]
Hi

I am modifying this [suggested code](https://stackoverflow.com/questions/75257369/pypdf2-paper-size-manipulation/75274841#75274841)

My actual code is:

```

from pypdf import PdfReader, PdfWriter, Transformation, PageObject, PaperSize
from pypdf.generic import RectangleObject

reader = PdfReader("input.pdf")
writer = PdfWriter()
for page in reader.pages:
  

  A4_w = PaperSize.A4.width
  A4_h = PaperSize.A4.height

  # resize page to fit *inside* A4
  h = float(page.mediabox.height)
  w = float(page.mediabox.width)
  scale_factor = min(A4_h/h, A4_w/w)

  transform = Transformation().scale(scale_factor,scale_factor).translate(0, A4_h/2 - h*scale_factor/2)
  page.add_transformation(transform)

  page.cropbox = RectangleObject((0, 0, A4_w, A4_h))

  # merge the pages to fit inside A4

  # prepare A4 blank page
  page_A4 = PageObject.create_blank_page(width = A4_w, height = A4_h)
  page.mediabox = page_A4.mediabox
  page_A4.merge_page(page)

  writer.add_page(page_A4)
writer.write('output.pdf')
```

input file: [link](https://drive.google.com/file/d/17RLTv1T74hh0maBPdIynmYTO-bLkkDAU/view?usp=drive_link) (767 KB)
output file: [link](https://drive.google.com/file/d/1fi2Slr9_6qjiscbxnU7w5GgGO1d1BGiN/view?usp=sharing) (2.4 MB)

The problem is that the file size increases drastically. This doesn't happen with all files. Only some. Mostly official documents.
Would this be a bug or am i doing something wrong ?

**Comment by pubpub-zz** [comment, marked as answer]
your content stream can be compress to reduce size:
https://pypdf.readthedocs.io/en/stable/user/file-size.html#lossless-compression

the output: 
[o1.pdf](https://github.com/user-attachments/files/15524526/o1.pdf) (<600 KB)

**Comment by zain910128** [comment]
That's great.
I'll try that.

But can you also tell me why the file size is increasing in the first place ?

**Reply by pubpub-zz** [reply]
the content in output is not compressed. This compression need to be called. Your pdf content lots of drawing the compression has a major impact

**Comment by zain910128** [comment]
Can you please elaborate on this part, i think i didn't fully understand:
"Your pdf content lots of drawing the compression has a major impact"

Thanks !

**Reply by pubpub-zz** [reply]
open your pdf with tools such as Pdfbox in debug mode and you will observe lots of drawing instructions.
all these are a text which make the file big. if some compression is applied this will reduce its size.

**Comment by zain910128** [comment]
Hi

I finally got around to trying this solution.
However, what I'm seeing is very strange. By using the compression function on each page, my file size is increasing.

Here's the code 
```
# this is the code with compression
def convert_to_A4_compressed(input_path, output_path):
    reader = PdfReader(input_path)
    writer = PdfWriter()
    for page in reader.pages:
        A4_w = PaperSize.A4.width
        A4_h = PaperSize.A4.height
        # resize page to fit *inside* A4
        h = float(page.mediabox.height)
        w = float(page.mediabox.width)
        scale_factor = min(A4_h / h, A4_w / w)
        page_A4 = writer.add_blank_page(width = A4_w, height = A4_h)
        page_A4.merge_transformed_page(
            page,
            Transformation().scale(scale_factor,scale_factor).translate(
                A4_w/2-w*scale_factor/2,
                A4_h/2-h*scale_factor/2)
            )
        page_A4.compress_content_streams() # <---- this is where I've made the change.
    writer.write(output_path)
```

When i use this code, the file size increases to 2.8MB
But if i remove that 2nd last line, then file size is 2.3MB

What's going on ?

I also noticed that if I run the compression separately, after the for loop has finished, then the file size does indeed decrease.

Please do let me know what's going on here.

Thanks !

**Reply by pubpub-zz** [reply]
most probably the page content has been uncompressed to plain text. try to apply recompression after processing:
https://pypdf.readthedocs.io/en/stable/user/file-size.html#lossless-compression

**Comment by Cyberware-tech** [comment]
I tried that also, but unfortunately, the file size still does not decrease.
I copied the code from your link even.

What results are you getting ?

My steps:
1. Use the file i provided earlier (767 KB)
2. Apple conversions and transformations as per the function I provided above (2.4MB)
3. Take output file and use the lossless compression method (3 MB)

**Reply by pubpub-zz** [reply]
can you please provide, input and output file

**Reply by Cyberware-tech** [reply]
Sure,
I actually provided input file in the first post in this discussion. But here it is again:
Input file: [link](https://drive.google.com/file/d/17RLTv1T74hh0maBPdIynmYTO-bLkkDAU/view?usp=drive_link) (767 KB)

Output after running following code:
[Step 2 output file](https://drive.google.com/file/d/1tloOnYNbvFP6RyAocUZl-uJEvhJTQjd6/view?usp=drive_link)
```
def convert_to_A4(input_path, output_path):
    reader = PdfReader(input_path)
    writer = PdfWriter()
    for page in reader.pages:
        A4_w = PaperSize.A4.width
        A4_h = PaperSize.A4.height
        # resize page to fit *inside* A4
        h = float(page.mediabox.height)
        w = float(page.mediabox.width)
        scale_factor = min(A4_h / h, A4_w / w)
        page_A4 = writer.add_blank_page(width = A4_w, height = A4_h)
        page_A4.merge_transformed_page(
            page,
            Transformation().scale(scale_factor,scale_factor).translate(
                A4_w/2-w*scale_factor/2,
                A4_h/2-h*scale_factor/2)
            )
    writer.write(output_path)
```

Output after compression, using the code
[Step 3 output file](https://drive.google.com/file/d/1EmRtrBHmqOoX3Yk_UM5w65KTSmnXpU-p/view?usp=drive_link)
```
from pypdf import PdfWriter
writer = PdfWriter(clone_from="output.pdf")
for page in writer.pages:
    page.compress_content_streams()  # This is CPU intensive!
with open("output_c2.pdf", "wb") as f:
    writer.write(f)
```

**Reply by Cyberware-tech** [reply]
PS: I just realised that I've used 2 accounts (OP and this one) to participate in this thread. Sorry for the confusion.

**Reply by pubpub-zz** [reply]
at first glance:
Steps2 contains uncompressed in an array.
in your second code you've created create  a new compressed object that replaces the array however the original object is left detached but within the object. 
This is the code I propose and where the output is less than 600 kB:
```
def convert_to_A4(input_path, output_path):
    from pypdf import PdfReader,PdfWriter,PaperSize,Transformation
    reader = PdfReader(input_path)
    writer = PdfWriter()
    for page in reader.pages:
        A4_w = PaperSize.A4.width
        A4_h = PaperSize.A4.height
        # resize page to fit *inside* A4
        h = float(page.mediabox.height)
        w = float(page.mediabox.width)
        scale_factor = min(A4_h / h, A4_w / w)
        page_A4 = writer.add_blank_page(width = A4_w, height = A4_h)
        page_A4.merge_transformed_page(
            page,
            Transformation().scale(scale_factor,scale_factor).translate(
                A4_w/2-w*scale_factor/2,
                A4_h/2-h*scale_factor/2)
            )
        page_A4.compress_content_streams() # to compress contents()
    from io import BytesIO;
    b = BytesIO()
    writer.write(b)
    b.seek(0)
    writer = PdfWriter(b)
    writer.write(output_path)
```

**Reply by Cyberware-tech** [reply]
Awesome
That works, Thanks a lot

But now I have another observation:

The result of your suggested code is less than 600kB
But the result of using the code example from the [documentation](https://pypdf.readthedocs.io/en/stable/user/file-size.html#lossless-compression) is greater than 600kB
Shouldn't both be similar size ? Is this difference expected or is the code in the documentation not correct ?

**Reply by pubpub-zz** [reply]
The writing through intermediate BytesIO() allows to get rid of unused object. This is not part of the example provided. this should explain the difference

**Reply by Cyberware-tech** [reply]
Ok, thanks for explaining.
It might be good to add this detail to the documentation, as it seems relevant.

**Reply by pubpub-zz** [reply]
Do not hesitate to propose a PR 😀
